fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+            data.pending.push(item);
+
+            // Start the thread if it hasn't already been started
+            let already_started = data.running;
+            data.running = true;
+            !already_started
+        };
+
+        if start_thread {
+            FLOG!(fd_monitor, "Thread starting");
+            let mut background_monitor = BackgroundFdMonitor {
+                data: Arc::clone(&self.data),
+                change_signaller: Arc::clone(&self.change_signaller),
+                items: Vec::new(),
+            };
+            crate::threads::spawn(move || {
+                background_monitor.run();
+            });
+        }
+
+        item_id
+    }
+
+    /// Avoid requiring a separate UniquePtr for each item C++ wants to add to the set by giving an
+    /// all-in-one entry point that can initialize the item on our end and insert it to the set.
+    fn add_item_ffi(&mut self, fd: RawFd, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> u64 {
+        // We need to convert between `const * c_void` to a `void_ptr` because we can only implement
+        // traits like `Send` and `Sync` for new types around raw pointers, but cxx bridge doesn't
+        // support passing the opaque type instead `*const c_void` pointer to our POD type.
+        // While we're at, we might as well transmute the correct types instead of using the
+        // equivalent of `void *` to eliminate mistakes on the rust side of things.
+        let callback = unsafe { std::mem::transmute(callback) };
+        let param = unsafe { std::mem::transmute(param) };
+        let mut item = FdMonitorItem::new();
+        item.fd.reset(fd);
+        item.callback = FdMonitorCallback::Ffi(callback, param);
+        if timeout_usecs != FdReadableSet::kNoTimeout {
+            item.timeout = Some(Duration::from_micros(timeout_usecs));
+        }
+        let item_id = self.add(item).0;
+        item_id
+    }
+
+    /// Mark that the item with the given ID needs to be woken up explicitly.
+    pub fn poke_item(&self, item_id: FdMonitorItemId) {
+        assert!(item_id.0 > 0, "Invalid item id!");
+        let needs_notification = {
+            let mut data = self.data.lock().expect("Mutex poisoned!");
+            let needs_notification = data.pokelist.is_empty();
+            // Insert it, sorted.
+            // TODO: The C++ code inserts it even if it's already in the poke list. That seems
+            // unnecessary?
+            let pos = match data.pokelist.binary_search(&item_id) {
+                Ok(pos) => pos,
+                Err(pos) => pos,
+            };
+            data.pokelist.insert(pos, item_id);
+            needs_notification
+        };
+
+        if needs_notification {
+            self.change_signaller.post();
+        }
+    }
+
+    fn poke_item_ffi(&self, item_id: u64) {
+        self.poke_item(FdMonitorItemId(item_id))
+    }
+
+    fn new_ffi() -> Box<Self> {
+        Box::new(Self {
```

Comment:
Most of this function wants to be reused by Rust once we port `io.cpp`.
Why not call this `new()` and move the `Box::new` wrapping to `new_fd_monitor_ffi()` (since the Box only exists for the CXX bridge)